dex/networks/dcr/script.go

```diff
@@ -120,216 +120,62 @@ const (
-	//
-	//  Notice that OP_0 is not specified since signature type 0 disabled.
-
-	if len(script) == 26 &&
-		script[0] == txscript.OP_DUP &&
-		script[1] == txscript.OP_HASH160 &&
-		script[2] == txscript.OP_DATA_20 &&
-		script[23] == txscript.OP_EQUALVERIFY &&
-		isStandardAltSignatureType(script[24]) &&
-		script[25] == txscript.OP_CHECKSIGALT {
-
-		return script[3:23], dcrec.SignatureType(asSmallInt(script[24]))
-	}
-
-	return nil, 0
-}
-
-// Check if the opcode is one of a small number of acceptable stake opcodes that
-// are prepended to P2PKH scripts. Return the opcode if it is, else OP_0.
-func stakeOpcode(script []byte) byte {
-	if len(script) == 0 {
-		return 0
-	}
-	opcode := script[0]
-	if opcode == txscript.OP_SSGEN || opcode == txscript.OP_SSRTX {
-		return opcode
-	}
-	return 0
-}
-
-// ExtractStakePubKeyHash extracts a pubkey hash from the passed public key
-// script if it is a standard pay-to-pubkey-hash script tagged with the provided
-// stake opcode.  It will return nil otherwise.
-func ExtractStakePubKeyHash(script []byte, stakeOpcode byte) []byte {
-	if len(script) == 26 &&
-		script[0] == stakeOpcode &&
-		script[1] == txscript.OP_DUP &&
-		script[2] == txscript.OP_HASH160 &&
-		script[3] == txscript.OP_DATA_20 &&
-		script[24] == txscript.OP_EQUALVERIFY &&
-		script[25] == txscript.OP_CHECKSIG {
-
-		return script[4:24]
-	}
-
-	return nil
-}
-
-// ExtractStakeScriptHash extracts a script hash from the passed public key
-// script if it is a standard pay-to-script-hash script tagged with the provided
-// stake opcode.  It will return nil otherwise.
-func ExtractStakeScriptHash(script []byte, stakeOpcode byte) []byte {
-	if len(script) == 24 &&
-		script[0] == stakeOpcode &&
-		script[1] == txscript.OP_HASH160 &&
-		script[2] == txscript.OP_DATA_20 &&
-		script[23] == txscript.OP_EQUAL {
-
-		return script[3:23]
+func convertScriptType(st stdscript.ScriptType) ScriptType {
+	var scriptType ScriptType
+	switch st {
+	// P2PKH
+	case stdscript.STPubKeyHashEcdsaSecp256k1:
+		scriptType = ScriptP2PKH
+	case stdscript.STPubKeyHashEd25519:
+		scriptType = ScriptP2PKH | ScriptSigEdwards
+	case stdscript.STPubKeyHashSchnorrSecp256k1:
+		scriptType = ScriptP2PKH | ScriptSigSchnorr
+	case stdscript.STStakeSubmissionPubKeyHash, stdscript.STStakeChangePubKeyHash,
+		stdscript.STStakeGenPubKeyHash, stdscript.STStakeRevocationPubKeyHash,
+		stdscript.STTreasuryGenPubKeyHash:
+		scriptType = ScriptP2PKH | ScriptStake
+	// P2SH
+	case stdscript.STScriptHash:
+		scriptType = ScriptP2SH
+	case stdscript.STStakeChangeScriptHash, stdscript.STStakeGenScriptHash,
+		stdscript.STStakeRevocationScriptHash, stdscript.STStakeSubmissionScriptHash,
+		stdscript.STTreasuryGenScriptHash:
+		scriptType = ScriptP2SH | ScriptStake
+	// P2MS
+	case stdscript.STMultiSig:
+		scriptType = ScriptMultiSig
+	default: // STNonStandard, STNullData, STTreasuryAdd, etc
+		return ScriptUnsupported
```

Comment:
The previous behavior was effectively to include in this `default` the stake outputs except for stakegen and stakerevocation.  We could go back to that, but I think we were still overlooking maturity requirements for both of these types, so my feeling now is that it's best to detect the script type as accurately as possible (within the bounds of our `ScriptType` enum) and determine spendability from the signer, dcrwallet, in the listunspent result.
Not totally set on that, but it feels right.